
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>smeagol package &#8212; SMEAGOL 0.0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="smeagol" href="modules.html" />
    <link rel="prev" title="Welcome to SMEAGOL’s documentation!" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="smeagol-package">
<h1>smeagol package<a class="headerlink" href="#smeagol-package" title="Permalink to this headline">¶</a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</section>
<section id="module-smeagol.aggregate">
<span id="smeagol-aggregate-module"></span><h2>smeagol.aggregate module<a class="headerlink" href="#module-smeagol.aggregate" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="smeagol.aggregate.collect_data">
<span class="sig-prename descclassname"><span class="pre">smeagol.aggregate.</span></span><span class="sig-name descname"><span class="pre">collect_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_pattern</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_with_index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_to_collect</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.aggregate.collect_data" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is used to aggregate multiple SMEAGOL result files, for example on different genomes. It collects a specific column (col_to_collect) from all the result files and merges these columns into a single dataframe. The merge is based on a specific index column (col_with_index) that should be present in all files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_pattern</strong> (<em>str</em>) – pattern to search for in filenames</p></li>
<li><p><strong>file_ids</strong> (<em>list</em>) – list of IDs, one per file</p></li>
<li><p><strong>col_with_index</strong> (<em>str</em>) – Name of the column to set as index</p></li>
<li><p><strong>col_to_collect</strong> (<em>str</em>) – Name of the column to collect</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A concatenated dataframe containing the <cite>col_to_collect</cite> column from all files matching the pattern <cite>file_pattern</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>df_all (pandas DataFrame)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-smeagol.encode">
<span id="smeagol-encode-module"></span><h2>smeagol.encode module<a class="headerlink" href="#module-smeagol.encode" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="smeagol.encode.SeqEncoding">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">smeagol.encode.</span></span><span class="sig-name descname"><span class="pre">SeqEncoding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">records</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rcomp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'none'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sense</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.encode.SeqEncoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>This class encodes a single nucleic acid sequence, or a set of sequences, all of which must have the same length. Genomic sequences used to initialize the class must also have the same sense.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="smeagol.encode.SeqEncoding.len">
<span class="sig-name descname"><span class="pre">len</span></span><a class="headerlink" href="#smeagol.encode.SeqEncoding.len" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of the sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="smeagol.encode.SeqEncoding.ids">
<span class="sig-name descname"><span class="pre">ids</span></span><a class="headerlink" href="#smeagol.encode.SeqEncoding.ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Numpy array containing the IDs of all sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="smeagol.encode.SeqEncoding.names">
<span class="sig-name descname"><span class="pre">names</span></span><a class="headerlink" href="#smeagol.encode.SeqEncoding.names" title="Permalink to this definition">¶</a></dt>
<dd><p>Numpy array containing the names of all sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="smeagol.encode.SeqEncoding.seqs">
<span class="sig-name descname"><span class="pre">seqs</span></span><a class="headerlink" href="#smeagol.encode.SeqEncoding.seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Numpy array containing the integer encoded sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="smeagol.encode.SeqEncoding.senses">
<span class="sig-name descname"><span class="pre">senses</span></span><a class="headerlink" href="#smeagol.encode.SeqEncoding.senses" title="Permalink to this definition">¶</a></dt>
<dd><p>Numpy array containing the senses (‘+’ or ‘-’) of all sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="smeagol.encode.SeqGroups">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">smeagol.encode.</span></span><span class="sig-name descname"><span class="pre">SeqGroups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">records</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rcomp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'none'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sense</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.encode.SeqGroups" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>This class encodes one or more groups of equal-length nucleic acid sequences.
Sequences in different groups may have different lengths.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.encode.integer_encode">
<span class="sig-prename descclassname"><span class="pre">smeagol.encode.</span></span><span class="sig-name descname"><span class="pre">integer_encode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.encode.integer_encode" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to encode a nucleic acid sequence as a sequence of integers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em><em>, </em><em>Seq</em><em> or </em><em>SeqRecord object</em>) – Nucleic acid sequence to encode. Allowed characters are A, C, G, T, U, Z, N, W, S, M, K, R, Y, B, D, H, V.</p></li>
<li><p><strong>rc</strong> (<em>bool</em>) – If True, reverse complement the sequence before encoding</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Numpy array containing the integer encoded sequence. The shape is (L, ) where L is the length of the sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>result (np.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.encode.one_hot_encode">
<span class="sig-prename descclassname"><span class="pre">smeagol.encode.</span></span><span class="sig-name descname"><span class="pre">one_hot_encode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.encode.one_hot_encode" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to one-hot encode a nucleic acid sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em><em>, </em><em>Seq</em><em> or </em><em>SeqRecord object</em>) – Nucleic acid sequence to encode. Allowed characters are A, C, G, T, U, Z, N, W, S, M, K, R, Y, B, D, H, V.</p></li>
<li><p><strong>rc</strong> (<em>bool</em>) – If True, reverse complement the sequence before encoding</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Numpy array containing the one-hot encoded sequence. The shape is (L, 4) where L is the length of the sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>result (np.array)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-smeagol.enrich">
<span id="smeagol-enrich-module"></span><h2>smeagol.enrich module<a class="headerlink" href="#module-smeagol.enrich" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="smeagol.enrich.enrich_in_genome">
<span class="sig-prename descclassname"><span class="pre">smeagol.enrich.</span></span><span class="sig-name descname"><span class="pre">enrich_in_genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">records</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rcomp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">simN</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">simK</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sense</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'+'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'binomial'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combine_seqs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_batch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shuf_seqs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.enrich.enrich_in_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to calculate enrichment of motifs in given sequence(s) relative to a background. The background sequences can be supplied directly or generated by shuffling the given sequence(s).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>records</strong> (<em>list</em>) – A list of seqrecord objects containing the sequence(s) to test. This list can be generated by reading a fasta file using ‘smeagol.io.read_fasta’.</p></li>
<li><p><strong>model</strong> (<a class="reference internal" href="#smeagol.models.PWMModel" title="smeagol.models.PWMModel"><em>PWMModel</em></a>) – An object of class PWMModel for PWM scanning</p></li>
<li><p><strong>rcomp</strong> (<em>str</em>) – ‘only’ to encode the reverse complements of the given sequences, ‘both’ to encode the reverse
complements as well as the given sequences, or ‘none’ to use only the given sequences without reverse complementing.</p></li>
<li><p><strong>threshold</strong> (<em>float</em>) – Threshold motif match score to use for PWM scanning. A sub-sequence that has a PWM match score higher than this will be counted as a match.</p></li>
<li><p><strong>simN</strong> (<em>int</em>) – The number of shuffled sequences to generate for the background model.</p></li>
<li><p><strong>simK</strong> (<em>int</em>) – The k-mer frequency to conserve while shuffling</p></li>
<li><p><strong>sense</strong> (<em>str</em>) – Sense of the given sequences. Must be ‘+’ or ‘-’</p></li>
<li><p><strong>background</strong> (<em>str</em>) – Background model for calculating p-values. Must be ‘binomial’ or ‘normal’</p></li>
<li><p><strong>combine_seqs</strong> (<em>bool</em>) – If true, combine outputs for all supplied sequences into a single output dataframe</p></li>
<li><p><strong>seq_batch</strong> (<em>int</em>) – The number of shuffled sequences to scan at a time. If 0, scan all at once.</p></li>
<li><p><strong>shuf_seqs</strong> (<em>str/list</em>) – A path to FASTA file or list of seqrecord objects containing shuffled sequences. Use if loading the background sequences directly instead of generating them by shuffling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dictionary containing results. Contains the following keys: ‘enrichment’ (enrichment statistics), ‘real_sites’ (locations of motif matches on the given sequences), ‘real_counts’ (number of sites for each PWM on the given sequences), ‘shuf_counts’ (number of sites for each PWM on the background sequences), ‘shuf_stats’ (mean and standard deviation of the number of sites for each PWM on the background sequences).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>results (dict)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.enrich.enrich_in_sliding_windows">
<span class="sig-prename descclassname"><span class="pre">smeagol.enrich.</span></span><span class="sig-name descname"><span class="pre">enrich_in_sliding_windows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alternative</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'two-sided'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.enrich.enrich_in_sliding_windows" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to test for enrichment of motifs in sliding windows across a genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sites</strong> (<em>pd.DataFrame</em>) – A pandas dataframe containing the locations of motifs in the genome. Should contain columns <cite>id</cite> (the sequence ID), <cite>start</cite> and <cite>end</cite>.</p></li>
<li><p><strong>genome</strong> (<em>list</em>) – A list of SeqRecord objects.</p></li>
<li><p><strong>width</strong> (<em>int</em>) – The width (in bases) of tiling windows to test.</p></li>
<li><p><strong>shift</strong> (<em>int</em>) – The shift (in bases) between successive windows. If not supplied, non-overlapping windows are generated.</p></li>
<li><p><strong>matrix_id</strong> (<em>str</em>) – Optional. The ID of the PWM to test. If supplied, it should correspond to an entry in the column <cite>matrix_id</cite> of <cite>sites</cite>.</p></li>
<li><p><strong>alternative</strong> (<em>str</em>) – Defines the alternative hypothesis. Options are ‘two-sided’, ‘less’ or ‘greater’. Default is ‘two-sided’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Pandas DataFrame containing the results of enrichment analysis in windows covering all sequences in genome</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.enrich.enrich_in_window">
<span class="sig-prename descclassname"><span class="pre">smeagol.enrich.</span></span><span class="sig-name descname"><span class="pre">enrich_in_window</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alternative</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'two-sided'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.enrich.enrich_in_window" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to test for enrichment of motifs in a subsequence relative to a complete sequence. This function uses Fisher’s exact test.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>window</strong> (<em>pd.DataFrame</em>) – A Pandas dataframe with the following columns: id (sequence ID), start, end</p></li>
<li><p><strong>sites</strong> (<em>pd.DataFrame</em>) – A pandas dataframe containing the locations of motifs in the genome. Should contain columns id (sequence ID, start, end.</p></li>
<li><p><strong>genome</strong> (<em>list</em>) – A list of SeqRecord objects</p></li>
<li><p><strong>matrix_id</strong> (<em>str</em>) – Optional. The ID of the selected PWM. If supplied, it must correspond to an entry in the column ‘Matrix_id’ of <cite>sites</cite>.</p></li>
<li><p><strong>alternative</strong> (<em>str</em>) – Defines the alternative hypothesis. Options are ‘two-sided’, ‘less’ or ‘greater’. Default is ‘two-sided’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe containing the result of enrichment analysis in selected window.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>result (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.enrich.examine_thresholds">
<span class="sig-prename descclassname"><span class="pre">smeagol.enrich.</span></span><span class="sig-name descname"><span class="pre">examine_thresholds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">records</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">simN</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">simK</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rcomp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'only'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold_shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sense</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'+'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combine_seqs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.enrich.examine_thresholds" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to scan given sequence(s) with PWMs and calculate the number of binding sites predicted using different cutoffs for the match score, for the real sequence as well as shuffled sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>records</strong> (<em>list</em>) – A list of seqrecord objects containing the sequence(s) to scan.</p></li>
<li><p><strong>model</strong> (<a class="reference internal" href="#smeagol.models.PWMModel" title="smeagol.models.PWMModel"><em>PWMModel</em></a>) – An object of class PWMModel for PWM scanning</p></li>
<li><p><strong>simN</strong> (<em>int</em>) – The number of shuffled sequences to generate for the background model. Default is 300.</p></li>
<li><p><strong>simK</strong> (<em>int</em>) – The k-mer frequency to conserve while shuffling. Default is 2.</p></li>
<li><p><strong>rcomp</strong> (<em>str</em>) – ‘only’ to encode the sequence reverse complements, ‘both’ to encode the reverse
complements as well as original sequences, or ‘none’. Default is ‘only’.</p></li>
<li><p><strong>min_threshold</strong> (<em>float</em>) – Minimum threshold motif match score to use for PWM scanning. Default is 0.5.</p></li>
<li><p><strong>threshold_shift</strong> (<em>float</em>) – Difference between successive thresholds to compare. Default is 0.1.</p></li>
<li><p><strong>sense</strong> (<em>str</em>) – Sense of the given sequences. Must be ‘+’ or ‘-’. Default is ‘+’.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – If true, output all information</p></li>
<li><p><strong>combine_seqs</strong> (<em>bool</em>) – If true, combine outputs for all supplied sequences into a single output dataframe</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dictionary containing results. Contains the following keys: ‘real_binned’ (Number of motif matches for each PWM in the given sequences, binned by their match score), ‘shuf_binned’ (Number of motif matches for each PWM in the background sequences, binned by their match score)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>results (dict)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-smeagol.io">
<span id="smeagol-io-module"></span><h2>smeagol.io module<a class="headerlink" href="#module-smeagol.io" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="smeagol.io.load_attract">
<span class="sig-prename descclassname"><span class="pre">smeagol.io.</span></span><span class="sig-name descname"><span class="pre">load_attract</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.io.load_attract" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to download all motifs and metadata from ATtRACT and load them as a pandas DF.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>contains matrices</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>df (pandas df)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.io.load_rbpdb">
<span class="sig-prename descclassname"><span class="pre">smeagol.io.</span></span><span class="sig-name descname"><span class="pre">load_rbpdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'human'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PWM'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.io.load_rbpdb" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Function to download all motifs and metadata from RBPDB and load them as a pandas DF.</dt><dd><p>Downloads version 1.3.1.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>species</strong> (<em>str</em>) – ‘human’, ‘mouse’, ‘fly’ or ‘worm’.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>contains matrices</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>df (pandas df)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.io.load_smeagol_PWMset">
<span class="sig-prename descclassname"><span class="pre">smeagol.io.</span></span><span class="sig-name descname"><span class="pre">load_smeagol_PWMset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'representative'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.io.load_smeagol_PWMset" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to download the curated set of motifs used in the SMEAGOL paper and load them as a pandas dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>dataset</strong> – ‘full’ or ‘representative’.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>contains matrices</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>df (pandas df)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.io.read_fasta">
<span class="sig-prename descclassname"><span class="pre">smeagol.io.</span></span><span class="sig-name descname"><span class="pre">read_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.io.read_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to read sequences from a fasta or fasta.gz file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file</strong> (<em>str</em>) – path to file</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of seqRecord objects, one for each sequence in the fasta file.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>records (list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.io.read_pms_from_dir">
<span class="sig-prename descclassname"><span class="pre">smeagol.io.</span></span><span class="sig-name descname"><span class="pre">read_pms_from_dir</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dirname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PPM'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transpose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.io.read_pms_from_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to read position matrices from a directory with separate files for each PM.</p>
<p>The input directory is expected to contain individual files each of which represents a
separate matrix.</p>
<p>The file name should be the name of the PWM followed by an extension. If <cite>_</cite> is included in
the name, the characters after the <cite>_</cite> will be dropped.</p>
<p>Matrices are by default expected to be in the position x base format, i.e. one row per position and
one column per base (A, C, G, T). If the matrices are instead in the base x position format, set
<cite>transpose=True</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dirname</strong> (<em>str</em>) – path to the folder containing PMs in individual files</p></li>
<li><p><strong>matrix_type</strong> (<em>str</em>) – ‘PWM’, ‘PPM’ (default) or ‘PFM’</p></li>
<li><p><strong>transpose</strong> (<em>bool</em>) – If true, transpose the matrix</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>pandas dataframe containing PMs</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.io.read_pms_from_file">
<span class="sig-prename descclassname"><span class="pre">smeagol.io.</span></span><span class="sig-name descname"><span class="pre">read_pms_from_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PPM'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_lens</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transpose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delimiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'\t'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.io.read_pms_from_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to read position matrices from a FASTA-like file.</p>
<p>The input file is expected to follow the format:</p>
<p>&gt;Matrix_1_ID
matrix values
&gt;Matrix_2_ID
matrix values</p>
<p>Matrices are by default expected to be in the position x base format, i.e. one row per position and
one column per base (A, C, G, T). If the matrices are instead in the base x position format, set
<cite>transpose=True</cite>.</p>
<p>Optionally, the ID rows may also contain a second field indicating the length of the
matrix, for example:</p>
<p>&gt;Matrix_1_ID&lt;tab&gt;7</p>
<p>The <cite>pwm.txt</cite> file downloaded from the Attract database follows this format. In this case, you can
set <cite>check_lens=True</cite> to confirm that the loaded PWMs match the expected lengths.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pm_file</strong> (<em>str</em>) – path to the file containing PMs</p></li>
<li><p><strong>matrix_type</strong> (<em>str</em>) – ‘PWM’, ‘PPM’ (default) or ‘PFM’</p></li>
<li><p><strong>check_lens</strong> (<em>bool</em>) – check that the matrix lengths match the lengths provided in the file</p></li>
<li><p><strong>transpose</strong> (<em>bool</em>) – transpose the matrices</p></li>
<li><p><strong>delimiter</strong> (<em>str</em>) – The string used to separate values in the file</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>pandas dataframe containing PMs</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.io.write_fasta">
<span class="sig-prename descclassname"><span class="pre">smeagol.io.</span></span><span class="sig-name descname"><span class="pre">write_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">records</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gz</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.io.write_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to write sequences to a fasta or fasta.gz file.</p>
<dl class="simple">
<dt>Params:</dt><dd><p>records (list): list of seqRecord objects to write.
file (str): path to file
gz (bool): If true, compress the output file with gzip.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Writes records to the file</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-smeagol.matrices">
<span id="smeagol-matrices-module"></span><h2>smeagol.matrices module<a class="headerlink" href="#module-smeagol.matrices" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.align_pms">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">align_pms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.align_pms" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to generate all possible overlaps between two position matrices.</p>
<dl class="simple">
<dt>Inputs:</dt><dd><p>X, Y (np.array): two position matrices.
min_overlap (int): minimum overlap allowed between matrices</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>tuples containing matrix start and end positions corresponding to each possible alignment of the two matrices.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>result</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.check_pfm">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">check_pfm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">freqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">warn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.check_pfm" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to check the validity of a PFM.
A valid PFM should have non-negative integer values and exactly 4 columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>freqs</strong> (<em>np.array</em>) – Numpy array containing frequency values. The array should be 2-dimensional and have 4 columns.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – if freqs is an invalid PFM.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.check_ppm">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">check_ppm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">probs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">warn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.check_ppm" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to check the validity of a PPM.
A valid PPM should have all values within the range [0,1], all rows summing to 1 (within a tolerance of eps), and exactly 4 columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>probs</strong> (<em>np.array</em>) – Numpy array containing probability values. The array should be 2-dimensional and have 4 columns.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – if probs is an invalid PPM.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.check_pwm">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">check_pwm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">weights</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.check_pwm" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to check the validity of a PWM.
A valid PWM should have exactly 4 columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>weights</strong> (<em>np.array</em>) – Numpy array containing PWM weights. The array should be 2-dimensional and have 4 columns.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – if weights is an invalid PWM.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.choose_cluster_representative_pms">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">choose_cluster_representative_pms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sims</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maximize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'median'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'weights'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PWM'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.choose_cluster_representative_pms" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to choose a representative position matrix from each cluster. This performs <cite>choose_representative_pm</cite> to choose a representative matrix within each supplied cluster.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas df</em>) – Dataframe containing position matrix values and IDs.</p></li>
<li><p><strong>sims</strong> (<em>np.array</em>) – pairwise similarities between all PWMs in pwms</p></li>
<li><p><strong>clusters</strong> (<em>list</em>) – cluster assignments for each PWM.</p></li>
<li><p><strong>maximize</strong> (<em>str</em>) – ‘mean’ or ‘median’. Metric  to choose representative matrix.</p></li>
<li><p><strong>weight_col</strong> (<em>str</em>) – column in pwms that contains matrix values.</p></li>
<li><p><strong>matrix_type</strong> (<em>str</em>) – PPM or PWM</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>IDs for the selected representative matrices.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>representatives (list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.choose_representative_pm">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">choose_representative_pm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sims</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maximize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'median'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'weights'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PWM'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.choose_representative_pm" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to choose a representative position matrix from a group. If 2 matrices are supplied, the one with lower entropy is chosen. For larger sets of matrices, the matrix with the highest mean (or median) normalized Pearson correlation to all other matrices is chosen.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas df</em>) – Dataframe containing position matrix values and IDs. The ID column should be named ‘Matrix_id’.</p></li>
<li><p><strong>sims</strong> (<em>np.array</em>) – pairwise similarities between all PWMs in df</p></li>
<li><p><strong>maximize</strong> (<em>str</em>) – ‘mean’ or ‘median’. Metric to choose representative matrix.</p></li>
<li><p><strong>weight_col</strong> (<em>str</em>) – the column in df that contains matrix values.</p></li>
<li><p><strong>matrix_type</strong> (<em>str</em>) – PPM or PWM</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>IDs for the selected representative matrices.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>result (list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.cluster_pms">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">cluster_pms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_clusters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sims</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'weights'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.cluster_pms" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to cluster position matrices.
A distance matrix between the matrices is computed using the normalized Pearson correlation metric and agglomerative clustering is used to find clusters. <cite>choose_representative_pm</cite> is called to identify a representative matrix from each cluster.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas df</em>) – Dataframe containing position matrix values and IDs.</p></li>
<li><p><strong>n_clusters</strong> (<em>int</em>) – Number of clusters</p></li>
<li><p><strong>sims</strong> (<em>np.array</em>) – pairwise similarities between all matrices in pwms</p></li>
<li><p><strong>weight_col</strong> (<em>str</em>) – column in pwms that contains matrix values.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>dictionary containing cluster labels, representative matrix IDs, and</dt><dd><p>minimum pairwise similarity within each cluster</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>result</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.cos_sim">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">cos_sim</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.cos_sim" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to calculate the cosine similarity between two vectors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<em>np.array</em>) – 1-D arrays.</p></li>
<li><p><strong>b</strong> (<em>np.array</em>) – 1-D arrays.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>cosine similarity between a and b.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>result (float)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.entropy">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">entropy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">probs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to calculate the entropy of a PPM or column of a PPM. Entropy is calculated using the formula <cite>H = -sum(p*log2(p))</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>probs</strong> (<em>np.array</em>) – Numpy array containing probability values.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Entropy value</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>result (float)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.matrix_correlation">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">matrix_correlation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.matrix_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to calculate the correlation between two equal-sized matrices. The values in each matrix are unrolled into a 1-D array and the Pearson correlation of the two 1-D arrays is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>np.array</em>) – two numpy arrays with same shape</p></li>
<li><p><strong>Y</strong> (<em>np.array</em>) – two numpy arrays with same shape</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pearson correlation between X and Y (unrolled into 1-D arrays)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>corr (float)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – if X and Y do not have equal shapes.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.ncorr">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">ncorr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.ncorr" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to calculate the normalized Pearson correlation between two position matrices, as defined in doi: 10.1093/nar/gkx314.</p>
<dl class="simple">
<dt>Inputs:</dt><dd><p>X, Y (np.array): two position matrices.
min_overlap (int): minimum overlap allowed between matrices</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>normalized Pearson correlation value</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>result (float)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.normalize_pm">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">normalize_pm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pm</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.normalize_pm" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to normalize a position matrix so that all rows sum to 1.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pm</strong> (<em>np.array</em>) – Numpy array containing probability values</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Numpy array with all rows normalized to sum to 1.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.order_pms">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">order_pms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.order_pms" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to order two matrices by width.</p>
<dl class="simple">
<dt>Inputs:</dt><dd><p>X, Y (np.array): two position matrices.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>two position matrices ordered by width, X being smaller.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>X, Y (np.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.pairwise_ncorrs">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">pairwise_ncorrs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mats</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.pairwise_ncorrs" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to calculate all pairwise normalized Pearson correlations between a list of position matrices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>mats</strong> (<em>list</em>) – a list of position matrices.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>All pairwise normalized Pearson correlations between matrices in mats.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ncorrs (np.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.pfm_to_ppm">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">pfm_to_ppm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">freqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocount</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.pfm_to_ppm" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to convert a valid PFM into a PPM. The matrix is normalized so that every position sums to 1, thus converting from a matrix of frequencies to a matrix of probabilities. To avoid zeros, a value of pseudocount/4 is first added to each position before normalization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>freqs</strong> (<em>np.array</em>) – array containing PFM values</p></li>
<li><p><strong>pseudocount</strong> (<em>float</em>) – pseudocount to add</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Numpy array containing PPM.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.position_wise_ic">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">position_wise_ic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">probs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.position_wise_ic" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to calculate the information content (IC) of each position in a PPM. The IC for a specific position is calculated using the formula: IC = 2 - sum(p*log2(p)), assuming a background probability of 0.25 for each of the 4 bases.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>probs</strong> (<em>np.array</em>) – Numpy array containing PPM probability values.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Numpy array containing the calculated information content of each column in probs.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>result (np.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.ppm_to_pwm">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">ppm_to_pwm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">probs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.ppm_to_pwm" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to convert a valid PPM into a PWM, using the formula: PWM = log2(PPM/B), where the background probability B is set to 0.25.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>probs</strong> (<em>np.array</em>) – Numpy array containing PPM probability values</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Numpy array containing the PWM. The shape of this array is (L, 4) where L is the PWM length.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.pwm_to_ppm">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">pwm_to_ppm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">weights</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.pwm_to_ppm" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.matrices.trim_ppm">
<span class="sig-prename descclassname"><span class="pre">smeagol.matrices.</span></span><span class="sig-name descname"><span class="pre">trim_ppm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">probs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frac_threshold</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.matrices.trim_ppm" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to trim non-informative positions from the ends of a PPM. See <cite>position_wise_ic</cite> for how information content of each position is calculated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>probs</strong> (<em>np.array</em>) – Numpy array containing PPM probability values</p></li>
<li><p><strong>frac_threshold</strong> (<em>float</em>) – threshold between 0 and 1. The mean information content of all columns in
the matrix will be calculated and any continuous columns in the beginning and start
of the matrix that have information content less than frac_threshold * mean IC will
be dropped.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Numpy array containing trimmed PPM.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>result (np.array)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-smeagol.models">
<span id="smeagol-models-module"></span><h2>smeagol.models module<a class="headerlink" href="#module-smeagol.models" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="smeagol.models.PWMModel">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">smeagol.models.</span></span><span class="sig-name descname"><span class="pre">PWMModel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pwm_df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.models.PWMModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class to contain a convolutional model used for PWM scanning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>Matrix_ids</strong> (<em>np.array</em>) – Numpy array containing the IDs of all the PWMs encoded in the model.</p></li>
<li><p><strong>widths</strong> (<em>np.array</em>) – Numpy array containing the widths of all the PWMs encoded in the model.</p></li>
<li><p><strong>max_scores</strong> (<em>np.array</em>) – Numpy array containing the maximum possible score for each PWM encoded in the model.</p></li>
<li><p><strong>channels</strong> (<em>int</em>) – Number of PWMs encoded in the model.</p></li>
<li><p><strong>max_width</strong> (<em>int</em>) – Width of the longest PWM encoded in the model.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="smeagol.models.PWMModel.set_model_weights">
<span class="sig-name descname"><span class="pre">set_model_weights</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.models.PWMModel.set_model_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode PWM weights in the model.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="smeagol.models.PWMModel.predict">
<span class="sig-name descname"><span class="pre">predict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.models.PWMModel.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Scan an integer encoded sequence with the model.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="smeagol.models.PWMModel.predict_with_threshold">
<span class="sig-name descname"><span class="pre">predict_with_threshold</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.models.PWMModel.predict_with_threshold" title="Permalink to this definition">¶</a></dt>
<dd><p>Scan an integer encoded sequence with the model and return matches with score above a specified threshold.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id0">
<span class="sig-name descname"><span class="pre">predict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id0" title="Permalink to this definition">¶</a></dt>
<dd><p>Scan encoded sequences with the model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seqs</strong> (<em>np.array</em>) – Numpy array containing integer encoded sequences.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id1">
<span class="sig-name descname"><span class="pre">predict_with_threshold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frac_threshold</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_batch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id1" title="Permalink to this definition">¶</a></dt>
<dd><p>Scan encoded sequences in batches with the model and threshold the returned values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<em>list</em><em> or </em><em>np.array</em>) – strings or integer encoded sequences</p></li>
<li><p><strong>frac_threshold</strong> (<em>float</em>) – fraction of maximum score to use as binding site detection threshold</p></li>
<li><p><strong>score</strong> (<em>bool</em>) – Output binding site scores as well as positions</p></li>
<li><p><strong>seq_batch</strong> (<em>int</em>) – number of sequences to scan at a time. If 0, scan all.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>positions where the input sequence(s) match the input</dt><dd><p>PWM(s) with a score above the specified threshold.</p>
</dd>
</dl>
<p>scores (np.array): score for each potential binding site</p>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>thresholded (np.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id2">
<span class="sig-name descname"><span class="pre">set_model_weights</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">weights</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id2" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to fix the weights of the conv model using PWM weights.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>weights</strong> (<em>np.array</em>) – Numpy array containing PWM weights.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-smeagol.scan">
<span id="smeagol-scan-module"></span><h2>smeagol.scan module<a class="headerlink" href="#module-smeagol.scan" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="smeagol.scan.count_in_sliding_windows">
<span class="sig-prename descclassname"><span class="pre">smeagol.scan.</span></span><span class="sig-name descname"><span class="pre">count_in_sliding_windows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.scan.count_in_sliding_windows" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to count binding sites in sliding windows across a genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sites</strong> (<em>pd.DataFrame</em>) – dataframe containing locations of binding sites</p></li>
<li><p><strong>genome</strong> (<em>list</em>) – list of SeqRecord objects</p></li>
<li><p><strong>matrix_id</strong> (<em>str</em>) – selected PWM</p></li>
<li><p><strong>width</strong> (<em>int</em>) – width of tiling windows</p></li>
<li><p><strong>shift</strong> (<em>int</em>) – shift between successive windows</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dataframe containing number of binding sites per window</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>results (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.scan.scan_sequences">
<span class="sig-prename descclassname"><span class="pre">smeagol.scan.</span></span><span class="sig-name descname"><span class="pre">scan_sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sense</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'+'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rcomp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'none'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outputs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['sites']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combine_seqs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_batch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.scan.scan_sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode given sequences and predict binding sites on them.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<em>list / str</em>) – list of seqrecord objects or fasta file</p></li>
<li><p><strong>model</strong> (<em>model</em>) – class PWMModel</p></li>
<li><p><strong>threshold</strong> (<em>float</em><em> or </em><em>np.arange</em>) – threshold (from 0 to 1) to identify binding sites OR np.arange (with binned_counts=True).</p></li>
<li><p><strong>sense</strong> (<em>str</em>) – sense of sequence(s), ‘+’ or ‘-‘.</p></li>
<li><p><strong>rcomp</strong> (<em>str</em>) – ‘only’ to encode the sequence reverse complement, ‘both’
to encode the reverse complement as well as original sequence, or ‘none’</p></li>
<li><p><strong>outputs</strong> (<em>list</em>) – List containing the desired outputs - any combination of ‘sites’, ‘counts’, ‘binned_counts’
and ‘stats’. For example: [‘sites’, ‘counts’].
‘sites’ outputs binding site locations
‘counts’ outputs the total count of binding sites per PWM
‘binned_counts’ outputs the counts of binding sites per PWM binned by score
‘stats’ outputs the mean and standard deviation of the number of binding sites per PWM,
across multiple sequences.</p></li>
<li><p><strong>score</strong> (<em>bool</em>) – output scores for binding sites. Only relevant if ‘sites’ is specified.</p></li>
<li><p><strong>group_by</strong> (<em>str</em>) – key by which to group sequences. If None, each sequence will be a separate group.</p></li>
<li><p><strong>combine_seqs</strong> (<em>bool</em>) – combine outputs for multiple sequence groups into a single dataframe</p></li>
<li><p><strong>sep_ids</strong> (<em>bool</em>) – separate outputs by sequence ID.</p></li>
<li><p><strong>seq_batch</strong> (<em>int</em>) – number of sequences to scan at a time. If 0, scan all.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dictionary containing specified outputs.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>output (dict)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.scan.score_site">
<span class="sig-prename descclassname"><span class="pre">smeagol.scan.</span></span><span class="sig-name descname"><span class="pre">score_site</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pwm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_wise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.scan.score_site" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to score a binding site sequence using a PWM.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pwm</strong> (<em>np.array</em>) – Numpy array containing the PWM weights</p></li>
<li><p><strong>seq</strong> (<em>str</em>) – A sequence the same length as the PWM.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>PWM match score</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>score (float)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-smeagol.utils">
<span id="smeagol-utils-module"></span><h2>smeagol.utils module<a class="headerlink" href="#module-smeagol.utils" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="smeagol.utils.get_Seq">
<span class="sig-prename descclassname"><span class="pre">smeagol.utils.</span></span><span class="sig-name descname"><span class="pre">get_Seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.utils.get_Seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to convert a sequence into a Seq object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seq</strong> (<em>str</em><em>, </em><em>Seq</em><em> or </em><em>SeqRecord</em>) – sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Seq object</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.utils.get_site_seq">
<span class="sig-prename descclassname"><span class="pre">smeagol.utils.</span></span><span class="sig-name descname"><span class="pre">get_site_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">site</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.utils.get_site_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to extract the sequence for a binding site.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>site</strong> – dataframe containing name, start and end.</p></li>
<li><p><strong>seqs</strong> (<em>list / str</em>) – list of seqrecord objects or fasta file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>sequence of binding site</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>result (str)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.utils.get_tiling_windows_over_genome">
<span class="sig-prename descclassname"><span class="pre">smeagol.utils.</span></span><span class="sig-name descname"><span class="pre">get_tiling_windows_over_genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.utils.get_tiling_windows_over_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to get tiling windows over a genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>list</em>) – list of SeqRecord objects</p></li>
<li><p><strong>width</strong> (<em>int</em>) – width of tiling windows</p></li>
<li><p><strong>shift</strong> (<em>int</em>) – shift between successive windows. By default the same as width.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>windows covering all sequences in genome</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>windows (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.utils.read_bg_seqs">
<span class="sig-prename descclassname"><span class="pre">smeagol.utils.</span></span><span class="sig-name descname"><span class="pre">read_bg_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">records</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">simN</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.utils.read_bg_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to read background sequences from FASTA file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>str</em>) – Path to fasta (or fasta.gz) file</p></li>
<li><p><strong>records</strong> (<em>list</em>) – Original un-shuffled records</p></li>
<li><p><strong>simN</strong> (<em>int</em>) – Number of shuffles</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of seqrecord objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.utils.shuffle_records">
<span class="sig-prename descclassname"><span class="pre">smeagol.utils.</span></span><span class="sig-name descname"><span class="pre">shuffle_records</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">records</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">simN</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">simK</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.utils.shuffle_records" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to shuffle sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>records</strong> (<em>list</em>) – list of seqRecord objects</p></li>
<li><p><strong>simN</strong> (<em>int</em>) – Number of times to shuffle</p></li>
<li><p><strong>simK</strong> (<em>int</em>) – k-mer frequency to conserve</p></li>
<li><p><strong>out_file</strong> (<em>str</em>) – Path to output file (optional)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of shuffled sequences
Writes shuf_records to file if out_file provided.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>shuf_records (list)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-smeagol.variant">
<span id="smeagol-variant-module"></span><h2>smeagol.variant module<a class="headerlink" href="#module-smeagol.variant" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="smeagol.variant.variant_effect_on_sites">
<span class="sig-prename descclassname"><span class="pre">smeagol.variant.</span></span><span class="sig-name descname"><span class="pre">variant_effect_on_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variants</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pwms</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.variant.variant_effect_on_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to predict variant effect on sites.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sites</strong> (<em>pd.DataFrame</em>) – dataframe containing information on each binding site to analyze. Contains columns ‘name’ (sequence name), ‘matrix_id’ (ID of the PWM that matched to the site), ‘start’ (site position) and ‘end’ (site end).</p></li>
<li><p><strong>variants</strong> (<em>pd.DataFrame</em>) – dataframe containing information on each variant to analyze. Contains
columns ‘name’ (sequence name), ‘pos’ (variant position), and ‘alt’ (alternate base at variant position).</p></li>
<li><p><strong>seqs</strong> (<em>list / str</em>) – list of seqrecord objects or fasta file.</p></li>
<li><p><strong>pwms</strong> (<em>pd.DataFrame</em>) – dataframe containing information on PWMs. Contains columns ‘matrix_id’ and ‘weights’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dataframe containing information on variant effects.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>effects (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-smeagol.visualize">
<span id="smeagol-visualize-module"></span><h2>smeagol.visualize module<a class="headerlink" href="#module-smeagol.visualize" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="smeagol.visualize.plot_background">
<span class="sig-prename descclassname"><span class="pre">smeagol.visualize.</span></span><span class="sig-name descname"><span class="pre">plot_background</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">enrichment_result</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Matrix_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sense</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(17,</span> <span class="pre">8)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.visualize.plot_background" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to plot the distribution of background counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>enrichment_result</strong> (<em>dict</em>) – Dictionary with result from enrich_in_genome</p></li>
<li><p><strong>Matrix_ids</strong> (<em>list</em>) – IDs of PWMs to plot</p></li>
<li><p><strong>sense</strong> (<em>str</em>) – ‘+’ or ‘-’</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em>) – total figure size</p></li>
<li><p><strong>ncols</strong> (<em>int</em>) – number of columns for figure panels</p></li>
<li><p><strong>file_path</strong> (<em>str</em>) – path to save figure.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Plot of motif distribution in real vs. background sequences.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.visualize.plot_binned_count_dist">
<span class="sig-prename descclassname"><span class="pre">smeagol.visualize.</span></span><span class="sig-name descname"><span class="pre">plot_binned_count_dist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">real_preds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Matrix_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sense</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shuf_preds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rounding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.visualize.plot_binned_count_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to plot distribution of (fractional) binding site scores in real vs. shuffled genomes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>real_preds</strong> (<em>pd.DataFrame</em>) – DF containing Matrix_id, sense, bin</p></li>
<li><p><strong>shuf_preds</strong> (<em>pd.DataFrame</em>) – DF containing Matrix_id, sense, bin</p></li>
<li><p><strong>Matrix_id</strong> (<em>str</em>) – ID of PWM for which to plot distribution</p></li>
<li><p><strong>sense</strong> (<em>str</em>) – sense of strand for which to plot distribution. If not given, both strands are used.</p></li>
<li><p><strong>file_path</strong> (<em>str</em>) – path to save figure</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Plot of binding site counts binned by score.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.visualize.plot_clustermap">
<span class="sig-prename descclassname"><span class="pre">smeagol.visualize.</span></span><span class="sig-name descname"><span class="pre">plot_clustermap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">foldchanges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pvalues</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row_cluster</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dendogram</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.visualize.plot_clustermap" title="Permalink to this definition">¶</a></dt>
<dd><p>Simple function to plot clustermap of foldchanges and pwms (without annotation); function will filter foldchanges whenever pvalues are significant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>foldchanges</strong> (<em>np.array</em>) – matrix of foldchanges for enrichment and depletion of pwms</p></li>
<li><p><strong>pvalues</strong> (<em>np.array</em>) – matrix of pvalues for enrichment of the same pwms</p></li>
<li><p><strong>threshold</strong> (<em>float</em>) – threshold for pvalues. default 0.05</p></li>
<li><p><strong>row_cluster</strong> (<em>bool</em>) – Cluster rows. Default True</p></li>
<li><p><strong>dendogram</strong> (<em>bool</em>) – Plot dendograms. Default True</p></li>
<li><p><strong>file_path</strong> (<em>str</em>) – The path to the file into which the plot should be written.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>heatmap plot.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.visualize.plot_clustermap_annot">
<span class="sig-prename descclassname"><span class="pre">smeagol.visualize.</span></span><span class="sig-name descname"><span class="pre">plot_clustermap_annot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">foldchanges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pvalues</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_zeros</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_occ</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_cluster</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row_cluster</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dendogram</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_dendogram</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row_dendogram</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.visualize.plot_clustermap_annot" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to plot clustermap of foldchanges and pwms, with annotation and more parameters</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>foldchanges</strong> (<em>np.array</em>) – matrix of foldchanges for enrichment and depletion of pwms</p></li>
<li><p><strong>pvalues</strong> (<em>np.array</em>) – matrix of pvalues for enrichment of the same pwms</p></li>
<li><p><strong>annot</strong> (<em>pd.DataFrame</em>) – annotation table with information on each column of pwm matrix</p></li>
<li><p><strong>column</strong> (<em>str</em>) – which column contains the name to be used as labels on the plot</p></li>
<li><p><strong>threshold</strong> (<em>float</em>) – threshold for pvalues. default 0.05</p></li>
<li><p><strong>group_by</strong> (<em>str</em>) – extra column in annot table that groups the occurrences (ex. Genus, Family, Kingdom, Host, etc) (Default: None). If no group_by function was provided but column parameter is “Species”, we try to cluster species by “Genus” using the first word of Species.</p></li>
<li><p><strong>drop_zeros</strong> (<em>bool</em>) – if True, drops all columns and rows in which all values are zero (Default: False)</p></li>
<li><p><strong>plot</strong> (<em>str</em>) – ‘sep’ if you wish to separate the plots according to the groups provided in group_by (Default: None)</p></li>
<li><p><strong>min_occ</strong> (<em>int</em>) – min number of occurrences within a group (defined by group_by) to plot separately if plot=”sep”</p></li>
<li><p><strong>cluster</strong> (<em>bool</em>) – Clusters both columns and rows (Default: True)</p></li>
<li><p><strong>col_cluster</strong> (<em>bool</em>) – Cluster columns (Default: True)</p></li>
<li><p><strong>row_cluster</strong> (<em>bool</em>) – Cluster rows (Default: True)</p></li>
<li><p><strong>dendogram</strong> (<em>bool</em>) – Plot both row and col dendograms (Default: True)</p></li>
<li><p><strong>row_dendogram</strong> (<em>bool</em>) – Plot row dendograms (Default: True)</p></li>
<li><p><strong>col_dendogram</strong> (<em>bool</em>) – Plot column dendograms (Default: True)</p></li>
<li><p><strong>file_path</strong> (<em>str</em>) – The path to the file into which the plot should be written.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>annotated heatmap plot.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.visualize.plot_ppm">
<span class="sig-prename descclassname"><span class="pre">smeagol.visualize.</span></span><span class="sig-name descname"><span class="pre">plot_ppm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ppm_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Matrix_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.visualize.plot_ppm" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to plot sequence logo from PPM</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ppm_df</strong> (<em>pd.DataFrame</em>) – Dataframe containing cols probs, Matrix_id</p></li>
<li><p><strong>Matrix_id</strong> – ID of PWM to plot</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Plots PPM</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.visualize.plot_pwm">
<span class="sig-prename descclassname"><span class="pre">smeagol.visualize.</span></span><span class="sig-name descname"><span class="pre">plot_pwm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pwm_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Matrix_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.visualize.plot_pwm" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to plot sequence logo from PWM</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pwm_df</strong> (<em>pd.DataFrame</em>) – Dataframe containing cols weight, Matrix_id</p></li>
<li><p><strong>Matrix_id</strong> – ID of PWM to plot (will be used as logo title)</p></li>
<li><p><strong>height</strong> (<em>int</em>) – Height of the plot</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Plots PWM</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.visualize.pwm_logo">
<span class="sig-prename descclassname"><span class="pre">smeagol.visualize.</span></span><span class="sig-name descname"><span class="pre">pwm_logo</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">weights</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.visualize.pwm_logo" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to visualize the sequence logo of a PWM.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>weights</strong> (<em>np.array</em>) – array containing PWM weight values</p></li>
<li><p><strong>title</strong> (<em>str</em>) – Title of the logo.</p></li>
<li><p><strong>height</strong> (<em>int</em>) – Height of the plot</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>PWM logo</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.visualize.sliding_window_count_plot">
<span class="sig-prename descclassname"><span class="pre">smeagol.visualize.</span></span><span class="sig-name descname"><span class="pre">sliding_window_count_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aspect</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.visualize.sliding_window_count_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to visualize site counts using sliding windows.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – A pandas dataframe that contains the data to plot.</p></li>
<li><p><strong>title</strong> (<em>str</em>) – The title to be used for the plot.</p></li>
<li><p><strong>cols</strong> (<em>int</em>) – number of columns for plot</p></li>
<li><p><strong>aspect</strong> (<em>float</em>) – width/height</p></li>
<li><p><strong>file_path</strong> (<em>str</em>) – The path to the file into which the plot should be written.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Sliding window plot.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="smeagol.visualize.sliding_window_enrichment_plot">
<span class="sig-prename descclassname"><span class="pre">smeagol.visualize.</span></span><span class="sig-name descname"><span class="pre">sliding_window_enrichment_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sliding_window_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_var</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_var</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xticklabels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(12,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#smeagol.visualize.sliding_window_enrichment_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to visualize enrichment / depletion analysis using sliding windows.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sliding_window_df</strong> (<em>pd.DataFrame</em>) – A pandas dataframe that contains the data to plot.</p></li>
<li><p><strong>x_var</strong> (<em>str</em>) – The name of the column that should be used as x-axis.</p></li>
<li><p><strong>y_var</strong> (<em>str</em>) – The name of the column that should be used as y-axis.</p></li>
<li><p><strong>xticklabels</strong> (<em>str</em>) – The name of the column that contains the x-axis labels.</p></li>
<li><p><strong>title</strong> (<em>str</em>) – The title to be used for the plot.</p></li>
<li><p><strong>file_path</strong> (<em>str</em>) – The path to the file into which the plot should be written.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Sliding window plot.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-smeagol">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-smeagol" title="Permalink to this headline">¶</a></h2>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">SMEAGOL</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">smeagol package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-smeagol.aggregate">smeagol.aggregate module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-smeagol.encode">smeagol.encode module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-smeagol.enrich">smeagol.enrich module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-smeagol.io">smeagol.io module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-smeagol.matrices">smeagol.matrices module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-smeagol.models">smeagol.models module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-smeagol.scan">smeagol.scan module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-smeagol.utils">smeagol.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-smeagol.variant">smeagol.variant module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-smeagol.visualize">smeagol.visualize module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-smeagol">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">smeagol</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">smeagol package</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Welcome to SMEAGOL’s documentation!</a></li>
      <li>Next: <a href="modules.html" title="next chapter">smeagol</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2022, Gruber Lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.5.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/smeagol.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>